kong/plugins/aws-lambda/handler.lua

```diff
@@ -25,32 +25,83 @@ function AWSLambdaHandler:new()
+  if content_type and string_find(content_type, "multipart/form-data", nil, true) then
     body_parameters = Multipart(ngx_req_get_body_data(), content_type):get_all()
-  elseif content_type and string_find(content_type:lower(), "application/json", nil, true) then
+  elseif content_type and string_find(content_type, "application/json", nil, true) then
     body_parameters, err = cjson.decode(ngx_req_get_body_data())
     if err then
       body_parameters = {}
     end
-  else
+  elseif content_type and string_find(content_type, "application/x-www-form-urlencoded", nil, true) then
     body_parameters = public_utils.get_post_args()
+  else
+    body_parameters = {}
   end
 
   return utils.table_merge(ngx_req_get_uri_args(), body_parameters)
 end
 
+local function retrieve_body(content_type)
+  local err, body
+  local raw_body = ngx_req_get_body_data()
+  if raw_body == nil then return nil end
+
+  if content_type and string_find(content_type, "application/json", nil, true) then
+    body, err = cjson.decode(raw_body)
+    if err then
+      body = {}
+    end
+  elseif content_type and ( string_find(content_type, "application/xml", nil, true)
+    or  string_find(content_type, "application/soap+xml", nil, true)
+    or  string_find(content_type, "text/", nil, true) )
+  then
+    body = raw_body
+  else
+    body = ngx_encode_base64(raw_body)
+  end
+  return body
+end
+
 function AWSLambdaHandler:access(conf)
   AWSLambdaHandler.super.access(self)
 
-  local bodyJson = cjson.encode(retrieve_parameters())
+  local var = ngx.var
+  local content_type = var.http_content_type
+  if content_type then
+     content_type = content_type:lower()
+  end
+
+  ngx_req_read_body()
+
+  local upstream_body
+
+  if conf.forward_request_body or conf.forward_http_headers
+     or conf.forward_http_method or conf.forward_request_uri then
+     upstream_body = {
+       request_body_args = not conf.forward_request_body and retrieve_parameters(content_type)
+                           or ngx_req_get_uri_args(),
+       request_body = conf.forward_request_body and retrieve_body(content_type) or nil,
+       request_http_headers = conf.forward_http_headers and ngx_req_get_headers() or nil,
+       request_http_method = conf.forward_http_method and var.request_method or nil,
+       request_uri = conf.forward_request_uri and var.request_uri or nil
+     }
+  else
+     upstream_body = retrieve_parameters(content_type)
+  end
+
+  local body_json, err = cjson.encode(upstream_body)
+  if not body_json then
+    ngx.log(ngx.ERR, "[aws-lambda] could not encode JSON ",
+                   "body for forwarded request values: ", err)
+  end
 
-  local host = string.format("lambda.%s.amazonaws.com", conf.aws_region)
-  local path = string.format("/2015-03-31/functions/%s/invocations",
+  local host = string_fmt("lambda.%s.amazonaws.com", conf.aws_region)
+  local path = string_fmt("/2015-03-31/functions/%s/invocations",
```

Comment:
This is another case where this should be done in a separate commit, outside this PR. Indeed this is the correct usage, but we want to keep commit atomic; since this change is unrelated, it should go in a separate PR.